core/domain/user_services_test.py

```diff
@@ -1318,6 +1329,398 @@ def test_record_user_started_state_translation_tutorial(self):
+        feedback_1: {}
+        hint_1: {}
+        rule_input_3: {}
+        solution: {}
+  New state:
+    classifier_model_id: null
+    content:
+      content_id: content
+      html: ''
+    interaction:
+      answer_groups: []
+      confirmed_unclassified_answers: []
+      customization_args: {}
+      default_outcome:
+        dest: New state
+        feedback:
+          content_id: default_outcome
+          html: ''
+        labelled_as_correct: false
+        missing_prerequisite_skill_id: null
+        param_changes: []
+        refresher_exploration_id: null
+      hints: []
+      id: null
+      solution: null
+    linked_skill_id: null
+    next_content_id_index: 0
+    param_changes: []
+    recorded_voiceovers:
+      voiceovers_mapping:
+        content: {}
+        default_outcome: {}
+    solicit_answer_details: false
+    card_is_checkpoint: false
+    written_translations:
+      translations_mapping:
+        content: {}
+        default_outcome: {}
+states_schema_version: 42
+tags: []
+title: Title
+""")
+
+    def setUp(self):
+        super(UserCheckpointProgressUpdateTests, self).setUp()
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+        self.signup(self.VIEWER_EMAIL, self.VIEWER_USERNAME)
+        self.owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+        self.viewer_id = self.get_user_id_from_email(self.VIEWER_EMAIL)
+
+        exp_services.save_new_exploration_from_yaml_and_assets(
+            self.owner_id, self.SAMPLE_EXPLORATION_YAML, self.EXP_ID, [])
+        self.exploration = exp_fetchers.get_exploration_by_id(self.EXP_ID)
+
+    def test_user_checkpoint_progress_is_updated_correctly(self):
+        self.login(self.VIEWER_EMAIL)
+        exploration_user_data = exp_fetchers.get_exploration_user_data(
+            self.viewer_id, self.EXP_ID)
+        self.assertIsNone(exploration_user_data)
+
+        # First checkpoint reached.
+        user_services.update_learner_checkpoint_progress(
+            self.viewer_id, self.EXP_ID, 'Introduction', 1)
+        exploration_user_data = exp_fetchers.get_exploration_user_data(
+            self.viewer_id, self.EXP_ID)
+        self.assertEqual(
+            exploration_user_data.furthest_reached_checkpoint_exp_version, 1)
+        self.assertEqual(
+            exploration_user_data.furthest_reached_checkpoint_state_name,
+            'Introduction')
+        self.assertEqual(
+            exploration_user_data.most_recently_reached_checkpoint_exp_version,
+            1)
+        self.assertEqual(
+            exploration_user_data.most_recently_reached_checkpoint_state_name,
+            'Introduction')
+
+        # Make 'New state' a checkpoint.
+        # Now version of the exploration becomes 2.
+        exp_services.update_exploration(
+            self.owner_id, self.EXP_ID, _get_change_list(
+                'New state',
+                exp_domain.STATE_PROPERTY_CARD_IS_CHECKPOINT,
+                True),
```

Comment:
Can you maybe put the change list into a separate variable and then put it here?

Reply:
ditto elsewhere

Reply:
Done!